internal/proxy/prometheus.go

```diff
@@ -0,0 +1,34 @@
+package proxy
+
+import (
+	"context"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+)
+
+type promURL string
+
+func NewPrometheus(prefix string, endpoint *url.URL) http.HandlerFunc {
+	var promKey promURL
+
+	director := func(r *http.Request) {
+		prometheusURL := r.Context().Value(promKey).(*url.URL)
+
+		r.URL.Scheme = prometheusURL.Scheme
+		r.URL.Host = prometheusURL.Host
+		// Normalize request so Prometheus doesn't know that request was proxied.
+		r.URL.Path = strings.TrimPrefix(r.URL.Path, prefix)
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Put the upstream PrometheusURL into the request's context for the director.
+		r = r.WithContext(context.WithValue(r.Context(), promKey, endpoint))
```

Comment:
Why do we  put this into context? Using context for hidden request meta should be avoided unless it's has to be there. This is because it's is easy to forgot to add it or clone it on the way and it is easy to overallocate by accident. What's the use case here?

This endpoint is definitely static, right?

Reply:
This is something I have copied over from one of our old iterations on an equivalent API. I have removed the authentication logic, and then I didn't put much thought on it. I'll clean it up and simplify.